### PR TITLE
Implement BufRead for Chain

### PR DESCRIPTION
**Pull request by troplin**
Addresses #32536

### CODE REVIEW

File: src/libstd/io/mod.rs [hunk 1]
```diff
@@ -1442,6 +1442,27 @@ impl<T: Read, U: Read> Read for Chain<T, U> {
     }
 }
 
+#[stable(feature = "chain_bufread", since = "1.9.0")]
+impl<T: BufRead, U: BufRead> BufRead for Chain<T, U> {
+    fn fill_buf(&mut self) -> Result<&[u8]> {
+        if !self.done_first {
+            match try!(self.first.fill_buf()) {
```
Comment (rphmeier) [inline src/libstd/io/mod.rs]: As I mentioned in #32536, you should probably use the `?` operator here.

**Comment by alexcrichton** [comment]
Thanks! Could you add some tests here as well?

**Comment by troplin** [comment]
@alexcrichton Where should I put the tests? `src/test/run-pass`?
Name it after the issue or after the feature?
How extensive do they have to be?

**Comment by alexcrichton** [comment]
Either `src/test/run-pass` or in `std::io` directly (there's a few others in there I believe). I'd like at least a smoke test to ensure that *anything* works but some corner case tests might also be useful.

**Comment by sfackler** [comment]
Looks good to me, but deferring the r to @alexcrichton 

(github unfortunately doesn't notify reviewers when new commits are pushed to a PR, so you may want to ping the comments after making changes, BTW)

**Comment by alexcrichton** [comment]
@bors: r+ f611e446624d288da7def4ad175405579c8bc310

Thanks @troplin!

**Comment by troplin** [comment]
Yay! :smiley:
Thanks everyone for the support!

**Pull request merged**